Add kernel-install plugin that calls ukify

keszybz opened the pull request:
Alternative for https://github.com/systemd/systemd/pull/27119.

Basic functionality works, and I'm making a draft PR for early comments. Signing is not done yet.

Also early microcode is not done at all. I think that https://github.com/systemd/systemd/pull/27111 is a better approach: let a separate plugin build a prefix initrd, instead of duplicating the logic in multiple places.

yuwata commented on src/ukify/ukify.py:
```diff
@@ -657,17 +657,19 @@ def parse_args(args=None):
         allow_abbrev=False,
         usage='''\
-usage: ukify [options…] linux initrd…
+usage: ukify [options…] LINUX INITRD…
```
As you can see in the commit message, `usage:` is doubled. Please drop it.

yuwata commented on src/kernel-install/meson.build:
```diff
@@ -1,11 +1,19 @@
 kernel_install_in = files('kernel-install.in')
-loaderentry_install_in = files('90-loaderentry.install.in')
+
+ukify_install = custom_target(
```
Maybe, better to conditionalize with `if want_ukify` ??

yuwata commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,192 @@
+except (KeyError, ValueError):
+    VERBOSE = False
+
+UKIFY = os.getenv('KERNEL_INSTALL_UKIFY', '/usr/lib/systemd/ukify')
```
Maybe, better to comment that the env var is for debugging or testing, and not set by kernel-install by default ?

poettering commented on src/ukify/ukify.py:
```diff
@@ -4,7 +4,7 @@
 # pylint: disable=consider-using-with,unspecified-encoding,line-too-long
 # pylint: disable=too-many-locals,too-many-statements,too-many-return-statements
-# pylint: disable=too-many-branches
+# pylint: disable=too-many-branches,fixme
```
hmm?

keszybz replied:
It disables a warning from pylint about 'TODO' and other comments. I use pylint to check syntax and common errors, and those warnings are annoying.

poettering commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,192 @@
+                cmdline = Path('/proc/cmdline')
+                filter = True
+                if not cmdline.exists():
+                    return None
```
not a fan of such ttocttou iterations. why not:

```
text = None
for f in ('/etc/kernel/cmdline', '/usr/lib/kernel/cmdline', '/proc/cmdline'):
    try:
        text = options.read_text()
    except FileNotFoundError:
        continue

options = text.split()
```

This isn't prone to ttocttou, and shorter too, and less indentation as well!

poettering commented:
There's a TODO list item you can drop:

```
* kernel-install:
  - optionally, support generating type #2 entries instead of type #1, including signing them
```

keszybz commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,192 @@
+    ]
+
+    log('+', shell_join(cmd))
+    subprocess.check_call(cmd)
```
Pffff. I did that. It was surprisingly hard and annoying. I'll push the conversion as a separate commit.

keszybz commented:
Updated. I still didn't get around to signing, but I want to see if the CI passes.
Changes:
- some tweaks that should make the tests pass
- adjustments to meson config
- `ukify` is imported, instead of being executed

yuwata commented on meson.build:
```diff
@@ -2393,7 +2393,9 @@ exe = executable(
                         versiondep],
```
Not a big matter, but the commit message contains `# Conflicts:` lines.

behrmann commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,203 @@
+
+    log(f'{output} has been created')
+
+def main():
+    opts = parse_args()
+    if opts.command != 'add':
+        return
+    if not we_are_wanted():
+        return
+
+    cmdline = kernel_cmdline()
+    if cmdline is None:
+        sys.exit('No cmdline configuration found')
+
+    make_uki(opts, cmdline)
```
Right now raised errors are never caught and will thus produce tracebacks if something goes wrong, the exit code will also ever be 1. 

Adding a custom error and wrapping the body of `main` in a try block, could be used to suppress that
```suggestion
class UkifyInstallError(Exception):
    def __init__(self, msg, rc=1):
        self.msg = msg
        self.rc = rc
    

def main():
    try:
        opts = parse_args()
        if opts.command != 'add':
            return
        if not we_are_wanted():
            return

        cmdline = kernel_cmdline()
        if cmdline is None:
            sys.exit('No cmdline configuration found')

        make_uki(opts, cmdline)
    except UkifyInstallError as e:
        print(e.msg, file=sys.stderr)
        sys.exit(e.rc)
```
This way some of the errors could then also get an exit code of 77 and `kernel_cmdline()` could throw the error instead of returning `None`. We would also see whether an exit was from an error we already anticipated or was a bug. This would also leave the door open to add some debug environment variable to always print the stack trace.

When going this route, the raises should look like, e.g.
```python
except ValueError as e:
    raise UkifyInstallError("Kernel options must not start with '@'", 77) from e
```

keszybz replied:
Yeah, I think it makes sense to suppress tracebacks. But I'm leaving that for later… As the discussion in mkosi shows, getting this just right is quite hard, and generally during early development its convenient to rather have too many tracebacks than too few.

behrmann commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,203 @@
+
+def log(*args, **kwargs):
+    if VERBOSE:
+        print(*args, **kwargs)
```
cpython's output is always buffered, maybe to be more shell like this should be
```suggestion
        print(*args, **kwargs, flush=True)
```

behrmann replied:
Yeah, I'm not sure it's strictly necessary, I just remember running into this problem in the past, but that was code using multi processing, so it had a lot of children. 

An alternative to the explicit flush in the function, would be using cpython's `-u` option 
> Force the stdout and stderr streams to be unbuffered.  This option has no effect on the stdin stream.

This might be a bit hard, since using `/usr/bin/env` this would need the `-S` option, which should be now be available everywhere, though I'm not sure about.

On that note, I noticed that this plugin uses `env`. The other kernel-install plugins use `/bin/sh`, e.g. force the default system shell. Should this plugin maybe also default to the system python path?

keszybz replied:
> Yeah, I'm not sure it's strictly necessary, I just remember running into this problem in the past, but that was code using multi processing, so it had a lot of children.
> 
> An alternative to the explicit flush in the function, would be using cpython's `-u` option
> 
> > Force the stdout and stderr streams to be unbuffered.  This option has no effect on the stdin stream.
> 
> This might be a bit hard, since using `/usr/bin/env` this would need the `-S` option, which should be now be available everywhere, though I'm not sure about.

Is there some problem that this is intended to solve? If not, I would keep default behaviour.
 
> On that note, I noticed that this plugin uses `env`. The other kernel-install plugins use `/bin/sh`, e.g. force the default system shell. Should this plugin maybe also default to the system python path?

We use `env` for python code. `/bin/bash` and `/bin/sh` are standarized locations, but the place where `python3` is installed varies. Some distros will replace the path in scripts to some desired system path. It is better leave it to them. For non-distro installations, using `env` is fine, it "just works".

behrmann commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,203 @@
+    if BOOT_STUB:
+        args += [f'--stub={BOOT_STUB}']
+
+    # Punish me harder.
+    # We want this:
+    #   ukify = importlib.machinery.SourceFileLoader('ukify', UKIFY).load_module()
+    # but it throws a DeprecationWarning.
+    # https://stackoverflow.com/questions/67631/how-can-i-import-a-module-dynamically-given-the-full-path
+    # https://github.com/python/cpython/issues/65635
+    # offer "explanations", but to actually load a python file without a .py extension,
+    # the "solution" is 4+ incomprehensible lines.
+    # The solution with runpy gives a dictionary, which isn't great, but will do.
+    ukify = runpy.run_path(UKIFY, run_name='ukify')
+    ukify['main'](args)
```
Don't be so hard on yourself, I actually think using `runpy` is a good and readable solution! :)

I do think though, that maybe this comment doesn't need to be shipped where everybody can read it. While I agree, that the one-liner is preferable and something like it should be added back to the stdlib, the alternative they provide as a recipe is not that inscrutable. The reason is that the Python devs wanted to split getting information about a module from loading it, which is why they added the module specs so now initialising a module is this three step process of getting the spec, making the module from the spec and then executing it in a clean namespace. Hooking the module up to `sys.modules` would probably not be necessary for ukify, since that seems to be needed to make relative imports work, but ukify is a single file.

That being said, I think I would have gone with your first solution of shelling out to `ukify`, as I don't really see an advantage in this dance, no matter which version.

keszybz replied:
The first comment on the stackoverflow question has ~~138~~139 upvotes, and it expresses my sentiment very well. I don't understand why Python developers are trying so hard to make this hard and not listening to people telling them that the new "solution" is not great.

Not shelling out has advantages. For the next version of this PR I reworked the ukify code to skip the option-parsing step. This avoids pitfalls with syntax and is generally nicer. It's certainly more work, but I hope it'll be worth it in the long run.

poettering commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,189 @@
+                ('/proc/cmdline',           True),
+        ]:
+            try:
+                options = Path(cmdline).read_text().split()
```
doesn't this lack a "break" somewhere? otherwise it will read all three files one after the other, and ignore the first two after reading them..

keszybz replied:
It does. Thanks for catching this.

But after looking at this again, I reworked this to have a separate function for reading of the file and changed the loop to only cover the first two cases, and have the third path in a separate chunk. It seems much more legible this way.

poettering commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,189 @@
+                ('/etc/kernel/cmdline',     False),
+                ('/usr/lib/kernel/cmdline', False),
+                ('/proc/cmdline',           True),
+        ]:
```
not a python guru, hence I was wondering: why use [] for the outer list and () for the inner lists?

keszybz replied:
For clarity: the paren characters are different so it's easier to see what is an inner paren and what is an outer paren.

poettering commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -158,17 +158,26 @@ def kernel_cmdline() -> str:
         f'--output={output}',
         f'--cmdline={cmdline}',
-    ]
+    ])
```
hmm, now you put together a cmdline you then parse again.

Wouldn't it be nicer to split parse_args() function in ukify.py in two and already pass a parsed parameter block into it?

keszybz replied:
Yes, it would be. This required a bit of a rework of ukify, but the next version will have that.

DaanDeMeyer commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,203 @@
+    return cmdline
+
+def make_uki(opts, cmdline):
+    output = opts.staging_area / 'uki.efi'
```
Can we be a bit smarter about the name? I'm not sure whether it should be in here or in kernel-install itself, but we have a bunch of logic in mkosi to determine the filename of the unified kernel image that I would love to be able to remove in the future (see install_unified_kernel() in mkosi)

keszybz replied:
This name is required by `90-uki-copy.install`. If we want to change this, we should change it in both places. (But I'm not sure we do. The file gets renamed later.)

keszybz replied:
I added a comment.

behrmann commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,203 @@
+
+def we_are_wanted() -> bool:
+    KERNEL_INSTALL_LAYOUT = os.getenv('KERNEL_INSTALL_LAYOUT')
+
```
I'm not sure that's the best choice, since there's a a few initrd generators out there, that already build UKIs but might not handle `KERNEL_INSTALL_UKI_GENERATOR` yet. 

Since I think most initrd generators have settled for the `50-` prefix and this is at `60-`, a better solution might be to default to ukify if the variable is unset and `uki.efi` does not exist in the staging directory.

keszybz replied:
If we want to change this, I think the right place would be `kernel-install`. Then all the plugins get a consistent view.

behrmann replied:
My point there was, that `kernel-install` cannot know whether anything down the road will be able to handle `KERNEL_INSTALL_UKI_GENERATOR`, so a fallback would have to be local to `60-ukify.install`, but thinking some more about this, I don't think this is worth it. This is something distros need to decide. Packagers can set the default for `KERNEL_INSTALL_UKI_GENERATOR` to whatever makes sense together with their initrd generator that might or might not handle building UKIs and tinkerers can be expected to add their own plugins for `kernel-install`.

keszybz commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,203 @@
+        description='kernel-install plugin to build a Unified Kernel Image',
+        allow_abbrev=False,
+        usage='''\
+usage: 60-ukify.install COMMAND KERNEL_VERSION ENTRY_DIR KERNEL_IMAGE INITRD…
```
I dropped --help from the usage. I guess it's enough if it's shown automatically.

(The reason for showing --help and --version is that those are alternative verbs really, i.e. when you call `prog --help`, all the other arguments which would be mandatory are ignored and may even not be present. But I guess everybody understands how --help works, we don't need to be pedantic in the syntax listing.)

poettering commented:
(btw, iirc dracut currently optoinally supports the rescue image concept, i.e. spitting out two images instead of one that are registered with the boot loader. Not sure this makes too much sense, but can we cover that somehow?)

behrmann commented:
> (btw, iirc dracut currently optoinally supports the rescue image concept, i.e. spitting out two images instead of one that are registered with the boot loader. Not sure this makes too much sense, but can we cover that somehow?)

This has come up before in discussions with Debian and Arch. I'm not entirely sold on the concept of a rescue image - I think boot counting and falling back to what booted previously, but generating multiple output for a single kernel version seems useful. Specifically for Arch this came up for mkinitcpio which has a concept of generating multiple profiles of configs (called presets) for a given kernel. This is currently not not a natural fit for kernel-install, but it would make kernel-install more useful to more people.

keszybz commented:
I'm leaving the question of rescue images on the side for now. I think it'd make sense to support this, because some people will want it. But I think it's best supported as a separate plugin that loads a separate configuration file. Since this is python, we can easily create a second module that imports this one, reconfigures a few paths and some other details, and then builds a second kernel. But let's leave that for later.

---

Update. I addressed the comments above, or I commented why not (or why not yet).

I found a very strange thing: test_ukify.py is executed in the CI, and reported as passing, but the tests were broken. I don't understand what's going on here. Let's see if the tests pass now.

ukify.py is now imported as a module and the relevant parts of the code are invoked as functions. There are some uglinesses here and there, but nothing too major.

A config file is supported for `ukify`, so pretty much everything can be configured for `60-ukify.install` via this config file.

I also added some more tests (and found some bugs on the way…).

poettering commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,225 @@
+        except FileNotFoundError:
+            continue
+
+    options = Path(cmdline).read_text().split()
```
hmm, not grokking this? to what is cmdline set here?

behrmann replied:
This probably should probably be
```suggestion
    options = Path("/proc/cmdline").read_text().split()
```

keszybz replied:
Arrrgh.

poettering commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,225 @@
+    ukify['apply_config'](opts2)
+    ukify['finalize_options'](opts2)
+    ukify['check_inputs'](opts2)
+    ukify['make_uki'](opts2)
```
btw, should these functions in ukify.py carry some comment or so that clarifies they are entrypoints from this external script? just so that people don't forget to update everything when they change it.

keszybz replied:
All functions that are public (no "_" prefix) are implicitly entry points… I think the CI should be enough: if one of those is ever renamed, we'll immediately get an error here. We distribute both files together, so we don't need cross-version compatibility.

behrmann commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,225 @@
+
+def kernel_cmdline_base() -> typing.Optional[typing.List[str]]:
+    if root := os.getenv('KERNEL_INSTALL_CONF_ROOT'):
+        return Path(root).joinpath('cmdline').read_text().split()
```
This will throw a `FileNotFoundError` if `$KERNEL_INSTALL_CONF_ROOT/cmdline` does not exist.

keszybz replied:
Yeah. The assumption is that if you use `KERNEL_INSTALL_CONF_ROOT`, then `cmdline` must be provided, because you're building a UKI that is independent of the host file system. I guess this could be changed, but it does make *some* sense.

behrmann replied:
I think that's reasonable, but I'm pretty sure that might be one of the errors that will be hit a bit more often, so I think catching it and printing a clear error would be helpful.

keszybz commented:
Rebased to fix conflict with the mkosi update in f997f91d7dcf4fe2c8ff4e73274e045e15ec71ea.
I also updated the `ukify(1)` in a new commit.
The new plugin also deserves a man page, but I didn't include it yet. I'll write it once the general approach of how the plugin should work is approved.

keszybz commented:
Rebased.

keszybz commented:
Rebased again.

yuwata commented:
Needs rebase again...

keszybz commented:
TEST-67-INTEGRITY:                  FAIL     (181 s), unrelated.

bluca reviewed (approved):
Maybe an explicit test case to enforce that the combination of config options and command line options works as expected would be useful

keszybz commented:
Updated with two changes:
- I added a blurb in the man page that cmdline options generally win, and that exceptions are documented,
- one more commit with a test for option overriding.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -653,162 +665,455 @@ def make_uki(opts):
+
+    for section_name, section in cp.items():
+        idx = section_name.find(':')
+        if idx > 0:
```
If there's no colon in the section name `idx` will be `-1`, is it intended that the section name can be empty (except for the colon)?

keszybz replied:
The code would support that, but I don't think this should be allowed. I'll add a check.

behrmann commented on src/kernel-install/60-ukify.install.in:
```diff
@@ -0,0 +1,224 @@
+        except FileNotFoundError:
+            continue
+
+    options = Path('/proc/cmdline').read_text().split()
```
Could `/proc` be missing, e.g. when kernel-install starts working on images? One can expect the others to be there, but an error message will be nicer than a stack trace.

keszybz replied:
kernel-install itself might support `--root`, but if plugins are invoked, they'll most likely be invoked with the API filesystems mounted. So it's quite unlikely that this script can be invoked without `/proc`.

behrmann commented on man/ukify.xml:
```diff
@@ -44,212 +44,302 @@
+    by <command>systemd-measure</command> is used. It is also possible to specify the
+    <varname>PCRPrivateKey=</varname>/<option>--pcr-private-key=</option>,
+    <varname>PCRPublicKey=</varname>/<option>--pcr-public-key=</option>, and
+    <varname>Phases=</varname>/<option>--phases=</option> arguments more than once. Signatures will be then
```
```suggestion
    <varname>Phases=</varname>/<option>--phases=</option> arguments more than once. Signatures will then be
```

behrmann commented on man/ukify.xml:
```diff
@@ -44,212 +44,302 @@
+          <term><varname>Linux=<replaceable>LINUX</replaceable></varname></term>
+          <term>positional argument <replaceable>LINUX</replaceable></term>
+
+          <listitem><para>A path to the vmlinux binary.</para></listitem>
```
Maybe 
```suggestion
          <listitem><para>A path to the kernel binary.</para></listitem>
```
since some distros call it `vmlinuz`?

keszybz commented:
Updated with the smaller changes requested. I didn't add the handling of errors. This will be quite a lot of work, and I'd like to get this version finally merged.

keszybz commented:
```
Processing triggers for initramfs-tools (0.140ubuntu13.1) ...
update-initramfs: Generating /boot/initrd.img-5.15.0-71-generic
No lz4 in /usr/bin:/sbin:/bin, using gzip
Errors were encountered while processing:
 udev
W: --force-yes is deprecated, use one of the options starting with --allow instead.
E: Sub-process /usr/bin/dpkg returned an error code (1)
autopkgtest [17:43:42]: ERROR: erroneous package: installation of basic binaries failed, exit code 100
blame: https://salsa.debian.org/systemd-team/systemd.git#upstream-ci
badpkg: installation of basic binaries failed, exit code 100
```
Whatever. Not related.
